pkg/ingester/active_series.go

```diff
@@ -0,0 +1,198 @@
+package ingester
+
+import (
+	"math"
+	"sync"
+	"time"
+
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"go.uber.org/atomic"
+
+	"github.com/cortexproject/cortex/pkg/ingester/client"
+	"github.com/cortexproject/cortex/pkg/util"
+)
+
+const (
+	numActiveSeriesStripes = 512
+)
+
+// ActiveSeries is keeping track of recently active series for a single tenant.
+type ActiveSeries struct {
+	stripes [numActiveSeriesStripes]activeSeriesStripe
+}
+
+// activeSeriesStripe holds a subset of the series timestamps for a single tenant.
+type activeSeriesStripe struct {
+	// Unix nanoseconds. Only used by purge. Zero = unknown.
+	// Updated in purge and when old timestamp is used when updating series (in this case, oldestEntryTs is updated
+	// without holding the lock -- hence the atomic).
+	oldestEntryTs atomic.Int64
+
+	mu     sync.RWMutex
+	refs   map[model.Fingerprint][]activeSeriesEntry
+	active int // Number of active entries in this stripe. Only decreased during purge.
+}
+
+// activeSeriesEntry holds a timestamp for single series.
+type activeSeriesEntry struct {
+	lbs   labels.Labels
+	nanos atomic.Int64 // Unix timestamp in nanoseconds.
+}
+
+func NewActiveSeries() *ActiveSeries {
+	c := &ActiveSeries{}
+
+	// Stripes are pre-allocated so that we only read on them and no lock is required.
+	for i := 0; i < numActiveSeriesStripes; i++ {
+		c.stripes[i].refs = map[model.Fingerprint][]activeSeriesEntry{}
+	}
+
+	return c
+}
+
+// Updates series timestamp to 'now'. Function is called to make a copy of labels if entry doesn't exist yet.
+func (c *ActiveSeries) UpdateSeries(series labels.Labels, now time.Time, labelsCopy func(labels.Labels) labels.Labels) {
+	fp := client.Fingerprint(series)
+	stripeID := util.HashFP(fp) % numActiveSeriesStripes
+
+	c.stripes[stripeID].updateSeriesTimestamp(now, series, fp, labelsCopy)
+}
+
+// Purge removes expired entries from the cache. This function should be called
+// periodically to avoid memory leaks.
+func (c *ActiveSeries) Purge(keepUntil time.Time) {
+	for s := 0; s < numActiveSeriesStripes; s++ {
+		c.stripes[s].purge(keepUntil)
+	}
+}
+
+func (c *ActiveSeries) Active() int {
+	total := 0
+	for s := 0; s < numActiveSeriesStripes; s++ {
+		total += c.stripes[s].getActive()
+	}
+	return total
+}
+
+func (s *activeSeriesStripe) updateSeriesTimestamp(now time.Time, series labels.Labels, fp model.Fingerprint, labelsCopy func(labels.Labels) labels.Labels) {
+	nowNanos := now.UnixNano()
+
+	e := s.findEntryForSeries(fp, series)
```

Comment:
I think there's a race condition: we get a reference of an `activeSeriesEntry` and then we do update it. In the meanwhile, the purge could remove it, because the read lock is kept only during `findEntryForSeries()`. However, the next remote write push will add it back, right? If so, LGTM, but I would like your opinion too.

Reply:
Well spotted. Should be fixed now. (There was another problem... since atomic integer was not a pointer, but part of "entry", which was stored in a slice as a plain struct, without pointer, it could actually *move* in memory when slice grows. That's why atomic integers are now pointers).